models/concerns/validate_data.rb

```diff
@@ -0,0 +1,250 @@
+    }
+  end
+
+  def ensure_no_conflicting_skills
+    all_skills = (@hash[:vehicles].to_a + @hash[:services].to_a + @hash[:shipments].to_a).flat_map{ |mission|
+      mission[:skills]
+    }.compact.uniq
+
+    return unless ['vehicle_partition_', 'work_day_partition_'].any?{ |str|
+      all_skills.any?{ |skill| skill.to_s.start_with?(str) }
+    }
+
+    raise OptimizerWrapper::UnsupportedProblemError.new(
+      "There are vehicles or services with 'vehicle_partition_*', 'work_day_partition_*' skills. " \
+      'These skill patterns are reserved for internal use and they would lead to unexpected behaviour.'
+    )
+  end
+
+  def check_matrices
+    # matrix_index consistency
+    if @hash[:matrices].nil? || @hash[:matrices].empty?
+      if @hash[:points]&.any?{ |p| p[:matrix_index] }
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          'There is a point with point[:matrix_index] defined but there is no matrix'
+        )
+      end
+    else
+      max_matrix_index = @hash[:points].max{ |p| p[:matrix_index] || -1 }[:matrix_index] || -1
+      matrix_not_big_enough = @hash[:matrices].any?{ |matrix_group|
+        Models::Matrix.field_names.any?{ |dimension|
+          matrix_group[dimension] &&
+            (matrix_group[dimension].size <= max_matrix_index ||
+              matrix_group[dimension].any?{ |col| col.size <= max_matrix_index })
+        }
+      }
+      if matrix_not_big_enough
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          'All matrices should have at least maximum(point[:matrix_index]) number of rows and columns'
+        )
+      end
+    end
+  end
+
+  def check_vehicles
+    @hash[:vehicles]&.each{ |v|
+      # vehicle time cost consistency
+      if v[:cost_waiting_time_multiplier].to_f > (v[:cost_time_multiplier] || 1)
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          'Cost_waiting_time_multiplier cannot be greater than cost_time_multiplier'
+        )
+      end
+
+      # matrix_id consistency
+      if v[:matrix_id] && (@hash[:matrices].nil? || @hash[:matrices].none?{ |m| m[:id] == v[:matrix_id] })
+        raise OptimizerWrapper::DiscordantProblemError.new('There is no matrix with id vehicle[:matrix_id]')
+      end
+    }
+  end
+
+  def check_services_and_shipments(configuration, schedule)
+    (@hash[:services].to_a + @hash[:shipments].to_a).each{ |mission|
+      if (mission[:minimum_lapse] || 0) > (mission[:maximum_lapse] || 2**32)
+        raise OptimizerWrapper::DiscordantProblemError.new('Minimum lapse can not be bigger than maximum lapse')
+      end
+
+      # TODO : this should be replaced next line when max_split does not use visits_number > 1 without schedule anymore
+      # next if schedule && schedule[:range_indices] || mission[:visits_number].to_i <= 1
+      next if configuration.nil? || schedule && schedule[:range_indices] || mission[:visits_number].to_i <= 1
```

Comment:
@braktar @senhalil I changed here : for now if we have no configuration (coming from max_split in general) then we do not do this check. 

PR is ready to make this check more general after we addit max split : https://github.com/Mapotempo/optimizer-api/pull/193